vllm/ex/ex.py

```diff
@@ -0,0 +1,245 @@
+
+    This backend works by partitioning the provided module into supported/unsupported
+    sub-modules.  The supported submodules are passed to an optimizer and then compiled
+    via an optional "final" backend.
+    """
+
+    # TODO: this probably needs additional context to avoid collisions, e.g.
+    # module/model name.
+    cc = CodeCache()
+
+    def __init__(self, backend: Optional[str] = 'inductor'):
+        self.backend = backend
+
+    def __call__(self, gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]) -> Callable:
+        # Must make a copy so that inductor backend doesn't choke.
+        gm = copy.copy(gm)
+
+        gm.graph.eliminate_dead_code()
+
+        logger.debug(f"Original module {gm}:\n{graph_print_tabular(gm.graph)}")
+        logger.debug(f"input_types: {[type(inp) for inp in example_inputs]}")
+
+        # TODO: store these in the root module state dictionary so that code for
+        # all sub-modules is shared?  Or should these be globals?
+        fgen = FusedOpGenerator()
+
+        ShapeProp(gm).propagate(*example_inputs)
+
+        part_gm, parts = partition_graph(gm, example_inputs)
+
+        logger.debug(f"Partitioned module: {part_gm.print_readable(False)}")
+        logger.debug(f"parts: {parts}")
+
+        # Get the current FakeTensorMode (there should be one since we are in
+        # a backend).
+        fake_mode = torch._guards.detect_fake_mode()
+
+        # There should be an existing fake_mode but double check.
+        assert fake_mode is not None
+
+        # Ensure that 'allow_non_fake_inputs' is True so that original
+        # 'example_inputs' can be used.
+        # Using unittest.mock is borrowed from torch/_guards.py
+        with unittest.mock.patch.object(fake_mode, "allow_non_fake_inputs", True):
+            # Determine example inputs for submodules.
+            # Note: static_shapes can be applied here if necessary.
+            mig = ModuleInputGenerator(part_gm, fake_mode)
+            mig.propagate(*example_inputs)
+
+            for name, m in part_gm.named_modules():
+                if module_in_partitions(parts, m):
+                    assert name in mig.module_args
+                    module_inputs = mig.module_args[name][0]
+
+                    # If one of the partitioned modules is called in multiple
+                    # places, we skip it.  This should not happen though.
+                    if not module_inputs:
+                        logger.debug(f"SKIPPING {name}: multiple callers.")
+                        continue
+
+                    logger.debug(f"Optimizing {name}.")
+                    m = optimize(backend_class.cc, fgen, m, module_inputs)
+                    setattr(part_gm, name, m)
+
+                    logger.debug(f"Optimized {name}: {m.print_readable(False)}")
+
+                    # TODO: don't really need to recompile if nothing was modified.
+                    m.recompile()
+
+                    if self.backend != None:
```

Comment:
What happens here if self.backend is None? Is that something we support?

Reply:
If the backend is None, it falls back to interpreting the python code as normal.  So it will just be running the rewritten code.  For now, I think this is what we want in most cases.